### PR TITLE
8324553: Shenandoah: Move periodic tasks closer to their collaborators

### PR DESCRIPTION
**Pull request by earthling-amzn**
This is a refactoring to move Shenandoah's periodic tasks from the control thread to their respective dependencies: the pacer and monitoring support. This change will facilitate reviews of the upcoming changes to isolate the generational mode changes.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324553](https://bugs.openjdk.org/browse/JDK-8324553): Shenandoah: Move periodic tasks closer to their collaborators (**Task** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - no project role) ⚠️ Review applies to [635633bb](https://git.openjdk.org/jdk/pull/17540/files/635633bbb923ab85ec59df4c21e154e1f9b46589)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17540/head:pull/17540` \
`$ git checkout pull/17540`

Update a local copy of the PR: \
`$ git checkout pull/17540` \
`$ git pull https://git.openjdk.org/jdk.git pull/17540/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17540`

View PR using the GUI difftool: \
`$ git pr show -t 17540`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17540.diff">https://git.openjdk.org/jdk/pull/17540.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17540#issuecomment-1906643763)

### CODE REVIEW

**Review by kdnilsen — APPROVED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahPacer.hpp [hunk 1]
```diff
@@ -29,8 +29,24 @@
...
+// Periodic task to notify blocked paced waiters.
+class ShenandoahPeriodicPacerNotify : public PeriodicTask {
+public:
+  explicit ShenandoahPeriodicPacerNotify(ShenandoahPacer* pacer) :
+    PeriodicTask(PeriodicTask::min_interval),
+    _pacer(pacer) { }
+
+  void task() override;
+private:
+  ShenandoahPacer* _pacer;
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahPacer.hpp:47]: Move this declaration at the beginning and make `_pacer` const, while you are at it?

File: src/hotspot/share/gc/shenandoah/shenandoahPacer.hpp [hunk 2]
```diff
@@ -48,6 +64,7 @@ class ShenandoahPacer : public CHeapObj<mtGC> {
   TruncatedSeq* _progress_history;
   Monitor* _wait_monitor;
   ShenandoahSharedFlag _need_notify_waiters;
+  ShenandoahPeriodicPacerNotify _notify_waiters;
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahPacer.hpp:67]: Let's call this `ShenandoahPeriodicPacerNotifyTask _notify_waiters_task`.

File: src/hotspot/share/gc/shenandoah/shenandoahMonitoringSupport.cpp [hunk 3]
```diff
@@ -103,3 +106,44 @@ void ShenandoahMonitoringSupport::update_counters() {
...
+
+void ShenandoahPeriodicCountersUpdate::notify_heap_changed() {
+  if (_do_counters_update.is_unset()) {
+    _do_counters_update.set();
+  }
+}
+
+void ShenandoahPeriodicCountersUpdate::set_forced_counters_update(bool value) {
+  _force_counters_update.set_cond(value);
+}
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahMonitoringSupport.cpp:149]: Newline is missing.

File: src/hotspot/share/gc/shenandoah/shenandoahMonitoringSupport.hpp [hunk 4]
```diff
@@ -44,14 +66,20 @@ class ShenandoahMonitoringSupport : public CHeapObj<mtGC> {
   HSpaceCounters* _space_counters;
 
   ShenandoahHeapRegionCounters* _heap_region_counters;
+  ShenandoahPeriodicCountersUpdate _counters_update;
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahMonitoringSupport.hpp:69]: `ShenandoahPeriodicCountersUpdateTask _counters_update_task;`?

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 5]
```diff
@@ -782,10 +780,26 @@ void ShenandoahHeap::op_uncommit(double shrink_before, size_t shrink_until) {
...
-    control_thread()->notify_heap_changed();
+    notify_heap_changed();
   }
 }
 
+void ShenandoahHeap::notify_heap_changed() {
+  // Update monitoring counters when we took a new region. This amortizes the
+  // update costs on slow path.
+  monitoring_support()->notify_heap_changed();
+
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:791]: Drop the new line.

**Review by shipilev — COMMENTED** [review]
I need to look at it with fresh eyes tomorrow, but a couple of stylistic comments now:

**Review by shipilev — APPROVED** [review]
Looks fine.

**Pull request closed without merging**